security/s2a/internal/crypter/halfconn.go

```diff
@@ -0,0 +1,165 @@
+package crypter
+
+import (
+	"fmt"
+	"golang.org/x/crypto/cryptobyte"
+	s2a_proto "google.golang.org/grpc/security/s2a/internal/proto"
+	"hash"
+	"sync"
+)
+
+// uint64Size is the size of a uint64 in bytes.
+const uint64Size = 8
+
+const (
+	tls13Key    = "tls13 key"
+	tls13Nonce  = "tls13 iv"
+	tls13Update = "tls13 traffic upd"
+)
+
+type S2AHalfConnection struct {
+	cs            ciphersuite
+	h             func() hash.Hash
+	aeadCrypter   s2aAeadCrypter
+	expander      hkdfExpander
+	seqCounter    counter
+	mutex         sync.Mutex
+	trafficSecret []byte
+	nonce         []byte
+	key           []byte
+}
+
+// NewHalfConn creates a new instance of S2AHalfConnection.
+func NewHalfConn(ciphersuite s2a_proto.Ciphersuite, trafficSecret []byte) (S2AHalfConnection, error) {
+	cs := newCiphersuite(ciphersuite)
+	if cs.trafficSecretSize() != len(trafficSecret) {
+		return S2AHalfConnection{}, fmt.Errorf("supplied traffic secret must be %v bytes, given: %v", cs.trafficSecretSize(), trafficSecret)
+	}
+
+	hc := S2AHalfConnection{cs: cs, h: cs.hashFunction(), expander: &defaultHKDFExpander{}, seqCounter: newCounter(), trafficSecret: trafficSecret}
+
+	var err error
+	if err = hc.updateWithNewTrafficSecret(hc.trafficSecret); err != nil {
+		return S2AHalfConnection{}, fmt.Errorf("hc.updateWithNewTrafficSecret(%v) failed with error: %v", hc.trafficSecret, err)
+	}
+
+	hc.aeadCrypter, err = cs.aeadCrypter(hc.key)
+	if err != nil {
+		return S2AHalfConnection{}, fmt.Errorf("cs.aeadCrypter(%v) failed with error: %v", hc.key, err)
```

Comment:
Please don't have the key in the logs. It would be better to say e.g. `Failed to initialize AEAD crypter.`